File: locust/purchase_checkout.py
from locust import HttpUser, between, task
from common import EMAIL, PASSWORD, PAYMENT_METHOD_ID, ProductionTrafficShape, SEARCH_QUERY, auth_headers, find_product


class PurchaseCheckout(HttpUser):
    wait_time = between(1, 3)

    @task
    def purchase(self):
        login = self.client.post("/api/auth/login", json={"email": EMAIL, "password": PASSWORD}, headers=auth_headers(), name="POST /api/auth/login")
        data = login.json() if login.status_code == 200 else {}
        token = data.get("tokens", {}).get("accessToken"); refresh = data.get("tokens", {}).get("refreshToken")
        search = self.client.get("/api/products/search", params={"q": SEARCH_QUERY}, headers=auth_headers(), name="GET /api/products/search")
        products = search.json().get("data", []) if search.status_code == 200 else []
        product_id = products[0].get("id") if products else find_product(self.client)
        if not product_id: return
        self.client.get(f"/api/products/{product_id}", headers=auth_headers(), name="GET /api/products/:id")
        self.client.post("/api/cart/items", json={"productId": product_id, "quantity": 1}, headers=auth_headers(token), name="POST /api/cart/items")
        self.client.get("/api/cart", headers=auth_headers(token), name="GET /api/cart")
        order = self.client.post("/api/orders", json={"items": [{"productId": product_id, "quantity": 1}], "shippingAddress": {"name": "Load User", "line1": "1 Performance Way", "city": "Bengaluru", "region": "KA", "postalCode": "560001", "country": "IN"}}, headers=auth_headers(token), name="POST /api/orders")
        order_id = order.json().get("data", {}).get("id") if order.status_code == 201 else None
        if order_id: self.client.post("/api/checkout/payment", json={"orderId": order_id, "paymentMethodId": PAYMENT_METHOD_ID}, headers=auth_headers(token), name="POST /api/checkout/payment")
        self.client.get("/api/orders", headers=auth_headers(token), name="GET /api/orders")
        if refresh: self.client.post("/api/auth/logout", json={"email": EMAIL, "password": PASSWORD, "refreshToken": refresh}, headers=auth_headers(), name="POST /api/auth/logout")


class PurchaseShape(ProductionTrafficShape):
    peak_rate = 29
    production_users = 905

File: k6/purchase-checkout.js
import http from 'k6/http';
import { check } from 'k6';
import { BASE_URL, EMAIL, PASSWORD, jsonParams, parseJson, scenarioOptions, think } from './common.js';

export const options = scenarioOptions(29, 905, 1810);

export default function () {
  let response = http.post(`${BASE_URL}/api/auth/login`, JSON.stringify({email: EMAIL, password: PASSWORD}), jsonParams());
  const auth = parseJson(response); const token = auth.tokens?.accessToken; const refreshToken = auth.tokens?.refreshToken;
  check(response, {'login 200': (r) => r.status === 200 && !!token}); think(1, 3);
  response = http.get(`${BASE_URL}/api/products/search?q=${encodeURIComponent(__ENV.SEARCH_QUERY || 'shoe')}`, jsonParams());
  let products = parseJson(response).data || [];
  if (!products.length) products = parseJson(http.get(`${BASE_URL}/api/products`, jsonParams())).data || [];
  const id = __ENV.PRODUCT_ID || products[0]?.id; check(response, {'search 200': (r) => r.status === 200}); think(1, 3);
  if (!id) return;
  check(http.get(`${BASE_URL}/api/products/${id}`, jsonParams()), {'detail 200': (r) => r.status === 200}); think(1, 3);
  response = http.post(`${BASE_URL}/api/cart/items`, JSON.stringify({productId: id, quantity: 1}), jsonParams(token)); check(response, {'add item 201': (r) => r.status === 201}); think(1, 3);
  check(http.get(`${BASE_URL}/api/cart`, jsonParams(token)), {'cart 200': (r) => r.status === 200}); think(1, 3);
  response = http.post(`${BASE_URL}/api/orders`, JSON.stringify({items: [{productId: id, quantity: 1}], shippingAddress: {name: 'Load User', line1: '1 Performance Way', city: 'Bengaluru', region: 'KA', postalCode: '560001', country: 'IN'}}), jsonParams(token));
  const orderId = parseJson(response).data?.id; check(response, {'order 201': (r) => r.status === 201}); think(1, 3);
  if (orderId) { response = http.post(`${BASE_URL}/api/checkout/payment`, JSON.stringify({orderId, paymentMethodId: __ENV.PAYMENT_METHOD_ID || 'pm_card_visa'}), jsonParams(token)); check(response, {'payment accepted': (r) => r.status === 200}); }
  think(1, 3); check(http.get(`${BASE_URL}/api/orders`, jsonParams(token)), {'orders 200': (r) => r.status === 200}); think(1, 3);
  check(http.post(`${BASE_URL}/api/auth/logout`, JSON.stringify({email: EMAIL, password: PASSWORD, refreshToken}), jsonParams()), {'logout 200': (r) => r.status === 200});
}
